### ISSUE TITLE
Don't use ReadOnlySettings at runtime

### ISSUE THREAD

**Issue by melkio** [Particular/NServiceBus.AzureServiceBus#282]
Taskforce: @danielmarbach @seanfeldman
## Original issue raised by @melkio

Transport shouldn't use `ReadOnlySettings` at "runtime". It should read settings and configure components passing raw values into constructor (or somewhere else).
This approach should be better for two reasons:
- avoid _boxing/unboxing_ with `ValueType`: `settings.Get<bool>(...)` invocation is casting `object` stored into dictionary to `bool`
- don't rely on a component (`ReadOnlySettings`) managed by Core, that could change behavior in the future

## Assumptions
1. The work of this taskforce is not going to be based on v7 release of the transport, but result in v7.x.
2. This is a starting issue and after retrospective, if found effective, spin up the next item.
3. The issue should solely focus on Settings and any other issues discovered during the work should be registered under ASB repo.

## Expectations
1. TF members to communicate to their maintainer groups and get them involved when needed, especially for discussions where GH issue is not the right medium to be used. 
2. Have mutual respect and understanding when dealing with code that is done the way it's done for historical reasons.

## PoA
- [X] Config
- [X] Creation already good
- [x] Connectivity https://github.com/Particular/NServiceBus.AzureServiceBus/pull/604
- [x] Receiving https://github.com/Particular/NServiceBus.AzureServiceBus/pull/605
- [x] Addressing https://github.com/Particular/NServiceBus.AzureServiceBus/pull/603
- [x] Sending #607
- [x] Topology #608

**Comment by yvesgoeleven** [comment]
I don't understand. Configuration API fills up this structure only, why should we not be allowed to use it? Are we supposed to copy everything over to some other structure?

**Comment by danielmarbach** [comment]
The settings API is a composition level API only. The design is the following:

After a certain point in the bootstrapping phase, the settings holder gets locked for modifications. Then the features get invoked. You get the settings holder as a segregated interface called `ReadOnlySettings`. Those settings are meant to be transferred to custom structure, bool flags, value types of any sorts. Like @melkio mentioned not following this design principle can have following consequences:

* If at some point core takes a decision to clear all settings after the feature configuration phase, your code breaks
* Using read-only settings during runtime does not make sense because
 1. The settings keys will never change after the feature configuration phase
 1. The settings are stored in a `ConcurrentDictionary<string, object>` and retrieved by looking for defaults or overrides. This means: A lookup is an expensive runtime operation which incurs significant boxing operations which lead to even more runtime overhead.

In summary using those settings during runtime is a loose / loose situation.

**Comment by yvesgoeleven** [comment]
So you suggest we read all settings that we may need from the ReadOnlySettings object once and store them somewhere else (class fields, or some other structure)

**Comment by danielmarbach** [comment]
We discussed this yesterday. Since the settings extension constructors are internal now we can switch the constructor to get an instance of a topology and then the topology itself can have strongly typed properties that represent settings for the topology and the subsystems

**Comment by SeanFeldman** [comment]
Remaining two (sending & topology) pending with PRs.

**Comment by SeanFeldman** [comment]
With all the refactorings in place, we can close this issue and @danielmarbach can retire. 
Well, maybe hold off a bit with umbrella drinks @danielmarbach 😃
